### PR TITLE
Fix SSL connection testing in Python 3.7

### PR DESCRIPTION
**Pull request by seanthegeek**
Closes issue #1549

---
This change is [Reviewable](https://reviewable.io/reviews/dpkp/kafka-python/1661)

### CODE REVIEW

File: kafka/conn.py [hunk 1]
```diff
@@ -357,6 +357,10 @@ def connect(self):
                 ret = self._sock.connect_ex(self._sock_addr)
             except socket.error as err:
                 ret = err.errno
+            except ValueError:
+                # Python 3.7 and higher rasies ValueError if a socket
```
Comment (jeffwidman) [inline kafka/conn.py]: typo: `rasies`

File: kafka/conn.py [hunk 2]
```diff
@@ -357,6 +357,10 @@ def connect(self):
                 ret = self._sock.connect_ex(self._sock_addr)
             except socket.error as err:
                 ret = err.errno
+            except ValueError:
```
Comment (jeffwidman) [inline kafka/conn.py]: Given that this is python 3.7 and above, and that `ValueError` is a common enough error, I'd like to minimize the chance of this swallowing something unexpected... 

So could you also add a check that we are running under python >= 3.7? And if not, re-raise and don't set `ret`

See here for details: https://stackoverflow.com/a/1093331/770425

**Comment by jeffwidman** [comment]
Thanks for this! A couple of comments... I have no way to test this, so which environment(s) have you tested it in?

File: kafka/conn.py (hunk 1 above)
Comment (seanthegeek) [reply kafka/conn.py]: Fixed. Thanks.

File: kafka/conn.py [hunk 3]
```diff
@@ -357,6 +357,13 @@ def connect(self):
                 ret = self._sock.connect_ex(self._sock_addr)
             except socket.error as err:
                 ret = err.errno
+            except ValueError as err:
+                # Python 3.7 and higher raises ValueError if a socket
+                # is already connected
+                if sys.version_info[0] >= 3 and sys.version_info[1] >= 7:
```
Comment (silentben) [inline kafka/conn.py]:
```suggestion
                if sys.version_info >= (3, 7):
```

**Comment by seanthegeek** [comment]
@jeffwidman Python 3.7 on Windows  and Python 3.6 on Debian Buster.

File: kafka/conn.py [hunk 4]
```diff
@@ -357,6 +357,13 @@ def connect(self):
                 ret = self._sock.connect_ex(self._sock_addr)
             except socket.error as err:
                 ret = err.errno
+            except ValueError as err:
+                # Python 3.7 and higher raises ValueError if a socket
+                # is already connected
+                if sys.version_info >= (3, 7):
+                    ret = None
+                else:
+                    raise err
```
Comment (jeffwidman) [inline kafka/conn.py]: IIRC, you only need this to be `raise` and it will re-raise the most recent one

File: build_integration.sh [hunk 5]
```diff
@@ -54,6 +54,9 @@ pushd servers
           fi
           echo
           echo "Extracting kafka ${kafka} binaries"
+          if [ ! -d ../$kafka ]; then
+            mkdir ../$kafka
+          fi
```
Comment (jeffwidman) [inline build_integration.sh]: Like above, this feels like a distinct PR / commit which wouldn't happen if I merge by squashing

File: .gitignore [hunk 6]
```diff
@@ -15,3 +15,4 @@ docs/_build
 integration-test/
 tests-env/
 .pytest_cache/
+venv/
```
Comment (jeffwidman) [inline .gitignore]: This makes sense to me to add, but as a separate commit from the SSL fix. 

However your commit history is a mess, so normally I'd merge this PR by squashing... but that would conflict with above...

**Comment by mooperd** [comment]
Hi,

With the following code:
```
from kafka import KafkaAdminClient

adminclient = KafkaAdminClient(
    bootstrap_servers="kafka-317b40d8-andrew-e213.aivencloud.com:18224",
    security_protocol="SSL",
    ssl_cafile="ca.pem",
    ssl_certfile="service.cert",
    ssl_keyfile="service.key",
)

adminclient.create_topics(['meow'], timeout_ms=None, validate_only=False)
```
I get an error:

```
Traceback (most recent call last):
  File "ensure-kafka-topic.py", line 10, in <module>
    ssl_keyfile="service.key",
  File "/Users/andrew/dev/aiven/kafka/admin/client.py", line 195, in __init__
    **self.config)
  File "/Users/andrew/dev/aiven/kafka/client_async.py", line 231, in __init__
    self.config['api_version'] = self.check_version(timeout=check_timeout)
  File "/Users/andrew/dev/aiven/kafka/client_async.py", line 857, in check_version
    version = conn.check_version(timeout=remaining, strict=strict, topics=list(self.config['bootstrap_topics_filter']))
  File "/Users/andrew/dev/aiven/kafka/conn.py", line 953, in check_version
    if not self.connect_blocking(timeout_at - time.time()):
  File "/Users/andrew/dev/aiven/kafka/conn.py", line 304, in connect_blocking
    self.connect()
  File "/Users/andrew/dev/aiven/kafka/conn.py", line 402, in connect
    if self._try_handshake():
  File "/Users/andrew/dev/aiven/kafka/conn.py", line 465, in _try_handshake
    self._sock.do_handshake()
  File "/Library/Frameworks/Python.framework/Versions/3.7/lib/python3.7/ssl.py", line 1112, in do_handshake
    self._check_connected()
  File "/Library/Frameworks/Python.framework/Versions/3.7/lib/python3.7/ssl.py", line 900, in _check_connected
    self.getpeername()
OSError: [Errno 57] Socket is not connected
```

Python 3.7.1 on OSX connecting to Aiven Kafka 2.0.1
System Version: macOS 10.13.5 (17F77)

**Comment by joagain** [comment]
Hi,

I'm getting a similar (but not exactly the same) error as @mooperd 's when calling `KafkaConsumer. beginning_offsets`:

```
  File "/Users/myuser/myproject/virtualenv/lib/python3.7/site-packages/kafka/consumer/group.py", line 986, in beginning_offsets
    partitions, self.config['request_timeout_ms'])
  File "/Users/myuser/myproject/virtualenv/lib/python3.7/site-packages/kafka/consumer/fetcher.py", line 213, in beginning_offsets
    partitions, OffsetResetStrategy.EARLIEST, timeout_ms)
  File "/Users/myuser/myproject/virtualenv/lib/python3.7/site-packages/kafka/consumer/fetcher.py", line 221, in beginning_or_end_offset
    offsets = self._retrieve_offsets(timestamps, timeout_ms)
  File "/Users/myuser/myproject/virtualenv/lib/python3.7/site-packages/kafka/consumer/fetcher.py", line 279, in _retrieve_offsets
    self._client.poll(future=future, timeout_ms=remaining_ms)
  File "/Users/myuser/myproject/virtualenv/lib/python3.7/site-packages/kafka/client_async.py", line 559, in poll
    self._maybe_connect(node_id)
  File "/Users/myuser/myproject/virtualenv/lib/python3.7/site-packages/kafka/client_async.py", line 386, in _maybe_connect
    conn.connect()
  File "/Users/myuser/myproject/virtualenv/lib/python3.7/site-packages/kafka/conn.py", line 402, in connect
    if self._try_handshake():
  File "/Users/myuser/myproject/virtualenv/lib/python3.7/site-packages/kafka/conn.py", line 465, in _try_handshake
    self._sock.do_handshake()
  File "/usr/local/Cellar/python/3.7.1/Frameworks/Python.framework/Versions/3.7/lib/python3.7/ssl.py", line 1112, in do_handshake
    self._check_connected()
  File "/usr/local/Cellar/python/3.7.1/Frameworks/Python.framework/Versions/3.7/lib/python3.7/ssl.py", line 900, in _check_connected
    self.getpeername()
OSError: [Errno 57] Socket is not connected
```

**Comment by seanthegeek** [comment]
Superseded by PR #1669

**Comment by jeffwidman** [comment]
@mooperd / @joagain I just merged https://github.com/dpkp/kafka-python/pull/1669.

Please try the latest `master` and see if that fixes these issues... if not, please open a new issue. @joagain in particular, based on a quick glance I don't think you should need to hack around that error, I think we should be handling at a library level. Although that is only based on a quick glance, so I may be missing something.

**Comment by danjo133** [comment]
I did: pip install git+https://github.com/dpkp/kafka-python after #1669 was merged and I need to add except OSError as well, otherwise I get the same:
OSError: [Errno 57] Socket is not connected

If I add the OSError catch there, it works for me (OSX, python 3.7.0)

**Comment by jeffwidman** [comment]
@danjo133 can you add this as a separate issue? Or better yet a new PR so that we can discuss the actual code change?

Commenting on old issues just tends to get lost in the noise...

**Pull request closed without merging**